File: src/scripts/inspect-patron-offers.ts
/**
 * Inspect why a specific patron gets "Light/Moderate/Strong" match labels.
 *
 * Usage:
 *   npx tsx src/scripts/inspect-patron-offers.ts P-000050
 */
import dotenv from "dotenv";
import { MongoClient } from "mongodb";

dotenv.config();

function cosineSimilarity(a: number[], b: number[]): number {
  const length = Math.min(a.length, b.length);
  if (length === 0) return 0;
  let dot = 0;
  let normA = 0;
  let normB = 0;
  for (let i = 0; i < length; i += 1) {
    dot += a[i] * b[i];
    normA += a[i] * a[i];
    normB += b[i] * b[i];
  }
  if (normA === 0 || normB === 0) return 0;
  return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}

function strengthLabel(score: number): string {
  if (score >= 0.75) return "Strong";
  if (score >= 0.5) return "Moderate";
  return "Light";
}

async function main() {
  const patronId = process.argv[2] ?? "P-000050";
  const uri = process.env.MONGODB_URI;
  const dbName = process.env.MONGODB_DB ?? "casino_marketing_demo";
  if (!uri) {
    console.error("MONGODB_URI is not set");
    process.exit(1);
  }

  const client = new MongoClient(uri);
  await client.connect();
  try {
    const db = client.db(dbName);
    const patron = await db
      .collection("patron_profiles")
      .findOne(
        { patronId },
        {
          projection: {
            _id: 0,
            patronId: 1,
            tier: 1,
            adt: 1,
            preferredGames: 1,
            pointsBalance: 1,
            preferenceEmbedding: 1,
          },
        }
      );

    if (!patron) {
      console.error(`Patron ${patronId} not found`);
      return;
    }

    const embedding = (patron.preferenceEmbedding ?? []) as number[];
    const nonZero = embedding.filter((x) => x !== 0).length;
    const norm = Math.sqrt(embedding.reduce((s, v) => s + v * v, 0));

    console.log("=".repeat(70));
    console.log(`Patron: ${patron.patronId}`);
    console.log(`  tier:           ${patron.tier}`);
    console.log(`  adt:            ${patron.adt}`);
    console.log(`  preferredGames: ${JSON.stringify(patron.preferredGames)}`);
    console.log(`  pointsBalance:  ${patron.pointsBalance}`);
    console.log(
      `  embedding:      dim=${embedding.length}, nonZero=${nonZero}, norm=${norm.toFixed(4)}`
    );
    if (norm < 1e-6) {
      console.log(
        "  ⚠ Embedding is effectively zero — scores will all be ~0 ⇒ all 'Light'."
      );
      console.log("    Run `npm run backfill` to generate real Voyage embeddings.");
    }

    const offers = await db
      .collection("offer_catalog")
      .find(
        {},
        {
          projection: {
            _id: 0,
            offerId: 1,
            title: 1,
            offerType: 1,
            targetGameTypes: 1,
            estimatedCost: 1,
            offerEmbedding: 1,
          },
        }
      )
      .toArray();

    const scored = offers
      .map((offer) => {
        const oEmb = (offer.offerEmbedding ?? []) as number[];
        const score = cosineSimilarity(embedding, oEmb);
        const oNonZero = oEmb.filter((x) => x !== 0).length;
        return {
          offerId: offer.offerId,
          title: offer.title,
          offerType: offer.offerType,
          targetGameTypes: offer.targetGameTypes,
          score,
          oNonZero,
        };
      })
      .sort((a, b) => b.score - a.score);

    console.log("\nTop 5 offers (by cosine similarity):");
    console.log("-".repeat(70));
    scored.slice(0, 5).forEach((row, i) => {
      console.log(
        `${i + 1}. ${row.offerId.padEnd(10)} ${strengthLabel(row.score).padEnd(8)} ` +
          `score=${row.score.toFixed(4)}  ${row.offerType.padEnd(20)} ${row.title}`
      );
      console.log(
        `   targetGames=${JSON.stringify(row.targetGameTypes)}  offerEmbedNonZero=${row.oNonZero}`
      );
    });

    const dist = {
      strong: scored.filter((s) => s.score >= 0.75).length,
      moderate: scored.filter((s) => s.score >= 0.5 && s.score < 0.75).length,
      light: scored.filter((s) => s.score < 0.5).length,
    };

    console.log("\nOverall match distribution across the catalog:");
    console.log(`  Strong   (≥0.75): ${dist.strong}`);
    console.log(`  Moderate (≥0.50): ${dist.moderate}`);
    console.log(`  Light    (<0.50): ${dist.light}`);
    console.log(`  Total offers:     ${scored.length}`);
    console.log("=".repeat(70));
  } finally {
    await client.close();
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: server/app/scripts/inspect_patron_offers.py
"""Inspect why a specific patron gets Light/Moderate/Strong offer matches.

Run with::

    uv run python -m app.scripts.inspect_patron_offers               # P-000050
    uv run python -m app.scripts.inspect_patron_offers P-000042      # custom
"""

from __future__ import annotations

import asyncio
import math
import sys

from app.collections import web_collections as cols
from app.db import close_client, get_db


def _cosine_similarity(a: list[float], b: list[float]) -> float:
    length = min(len(a), len(b))
    if length == 0:
        return 0.0
    dot = 0.0
    norm_a = 0.0
    norm_b = 0.0
    for i in range(length):
        dot += a[i] * b[i]
        norm_a += a[i] * a[i]
        norm_b += b[i] * b[i]
    if norm_a == 0 or norm_b == 0:
        return 0.0
    return dot / (math.sqrt(norm_a) * math.sqrt(norm_b))


def _strength_label(score: float) -> str:
    if score >= 0.75:
        return "Strong"
    if score >= 0.5:
        return "Moderate"
    return "Light"


async def main() -> None:
    patron_id = sys.argv[1] if len(sys.argv) > 1 else "P-000050"
    db = get_db()

    patron = await db[cols.patrons].find_one(
        {"patronId": patron_id},
        projection={
            "_id": 0,
            "patronId": 1,
            "tier": 1,
            "adt": 1,
            "preferredGames": 1,
            "pointsBalance": 1,
            "preferenceEmbedding": 1,
        },
    )

    if patron is None:
        print(f"Patron {patron_id} not found", file=sys.stderr)
        return

    embedding: list[float] = patron.get("preferenceEmbedding") or []
    non_zero = sum(1 for x in embedding if x != 0)
    norm = math.sqrt(sum(v * v for v in embedding))

    bar = "=" * 70
    print(bar)
    print(f"Patron: {patron.get('patronId')}")
    print(f"  tier:           {patron.get('tier')}")
    print(f"  adt:            {patron.get('adt')}")
    print(f"  preferredGames: {patron.get('preferredGames')}")
    print(f"  pointsBalance:  {patron.get('pointsBalance')}")
    print(
        f"  embedding:      dim={len(embedding)}, nonZero={non_zero}, norm={norm:.4f}"
    )
    if norm < 1e-6:
        print(
            "  ⚠ Embedding is effectively zero — scores will all be ~0 ⇒ all 'Light'."
        )
        print(
            "    Run `uv run python -m app.scripts.backfill_embeddings` to generate "
            "real Voyage embeddings."
        )

    offers = await (
        db[cols.offers]
        .find(
            {},
            projection={
                "_id": 0,
                "offerId": 1,
                "title": 1,
                "offerType": 1,
                "targetGameTypes": 1,
                "estimatedCost": 1,
                "offerEmbedding": 1,
            },
        )
        .to_list(length=None)
    )

    scored = []
    for offer in offers:
        o_emb: list[float] = offer.get("offerEmbedding") or []
        score = _cosine_similarity(embedding, o_emb)
        o_non_zero = sum(1 for x in o_emb if x != 0)
        scored.append(
            {
                "offerId": offer.get("offerId"),
                "title": offer.get("title"),
                "offerType": offer.get("offerType"),
                "targetGameTypes": offer.get("targetGameTypes"),
                "score": score,
                "oNonZero": o_non_zero,
            }
        )
    scored.sort(key=lambda s: s["score"], reverse=True)

    print("\nTop 5 offers (by cosine similarity):")
    print("-" * 70)
    for i, row in enumerate(scored[:5]):
        print(
            f"{i + 1}. {str(row['offerId']):<10} {_strength_label(row['score']):<8} "
            f"score={row['score']:.4f}  {str(row['offerType']):<20} {row['title']}"
        )
        print(
            f"   targetGames={row['targetGameTypes']}  "
            f"offerEmbedNonZero={row['oNonZero']}"
        )

    dist = {
        "strong": sum(1 for s in scored if s["score"] >= 0.75),
        "moderate": sum(1 for s in scored if 0.5 <= s["score"] < 0.75),
        "light": sum(1 for s in scored if s["score"] < 0.5),
    }

    print("\nOverall match distribution across the catalog:")
    print(f"  Strong   (≥0.75): {dist['strong']}")
    print(f"  Moderate (≥0.50): {dist['moderate']}")
    print(f"  Light    (<0.50): {dist['light']}")
    print(f"  Total offers:     {len(scored)}")
    print(bar)


async def _entrypoint() -> None:
    try:
        await main()
    finally:
        await close_client()


if __name__ == "__main__":
    asyncio.run(_entrypoint())
